server/plugin.go

```diff
@@ -599,22 +670,84 @@ func (p *Plugin) HasGroupHook(ctx context.Context, user *gitlab.UserInfo, namesp
 	return found, err
 }
 
-func (p *Plugin) checkAndRefreshToken(token *oauth2.Token) (*oauth2.Token, error) {
-	// If there is only one minute left for the token to expire, we are refreshing the token.
-	// The detailed reason for this can be found here: https://github.com/golang/oauth2/issues/84#issuecomment-831492464
-	// We don't want the token to expire between the time when we decide that the old token is valid
-	// and the time at which we create the request. We are handling that by not letting the token expire.
-	if time.Until(token.Expiry) <= 1*time.Minute {
-		conf := p.getOAuthConfig()
-		src := conf.TokenSource(context.Background(), token)
-		newToken, err := src.Token() // this actually goes and renews the tokens
-		if err != nil {
-			return nil, errors.Wrap(err, "unable to get the new refreshed token")
+func (p *Plugin) refreshToken(userInfo *gitlab.UserInfo, token *oauth2.Token) (*oauth2.Token, error) {
+	conf := p.getOAuthConfig()
+	src := conf.TokenSource(context.Background(), token)
+
+	newToken, err := src.Token() // this actually goes and renews the tokens
+
+	if err != nil {
+		if strings.Contains(err.Error(), "\"error\":\"invalid_grant\"") {
+			p.handleRevokedToken(userInfo)
 		}
-		if newToken.AccessToken != token.AccessToken {
-			return newToken, nil
+		return nil, errors.Wrap(err, "unable to get the new refreshed token")
+	}
+
+	if newToken.AccessToken != token.AccessToken {
+		p.API.LogDebug("Gitlab token refreshed.", "UserID", userInfo.UserID, "Gitlab Username", userInfo.GitlabUsername)
+		unencryptedNewToken := newToken.AccessToken // needed because the storeGitlabUserInfo method changes its value to an encrypted value
+
+		if err := p.storeGitlabUserToken(userInfo.UserID, newToken); err != nil {
+			return nil, errors.Wrap(err, "unable to store user info with refreshed token")
 		}
+
+		newToken.AccessToken = unencryptedNewToken
+		return newToken, nil
+	}
+
+	return token, nil
+}
+
+func (p *Plugin) handleRevokedToken(info *gitlab.UserInfo) {
+	p.disconnectGitlabAccount(info.UserID)
+	err := p.CreateBotDMPost(info.UserID, "Your GitLab account was disconnected due to an invalid or revoked authorization token. Reconnect your account using the `/gitlab connect` command.", "custom_git_revoked_token")
```

Comment:
As we discussed offline, maybe we should check if the disconnect was successful before sending this message. This way we don't spam the user with these messages on concurrent attempts to access the API.

@hanzei Note that this only occurs when the token was literally revoked externally, e.g. from Gitlab's UI

Reply:
And making this change would require another use of a mutex, correct?

Reply:
@mickmister Haven't implemented this change, waiting on reviewers' decision on if it's necessary.

Reply:
Surely, `disconnectGitlabAccount` could return an `error` (maybe even a multierror), but this seems like a nit pick to me.

Reply:
I just left it that way since it didn't return an error to begin with and it didn't seem particularly relevant to this PR to change that behavior.